tilequeue/metatile.py

```diff
@@ -5,57 +5,112 @@
+            # of coord.
+            delta_z = coord.zoom - parent.zoom
+            assert delta_z >= 0, "Coordinates must be descendents of parent"
+
+            # change in row/col coordinates are relative to the upper left
+            # coordinate at that zoom. both should be positive.
+            delta_row = coord.row - (int(parent.row) << delta_z)
+            delta_column = coord.column - (int(parent.column) << delta_z)
+            assert delta_row >= 0, \
+                "Coordinates must be contained by their parent, but " + \
+                "row is not."
+            assert delta_column >= 0, \
+                "Coordinates must be contained by their parent, but " + \
+                "column is not."
+
+            tile_name = '%d/%d/%d.%s' % \
+                (delta_z, delta_column, delta_row, tile['format'].extension)
             tile_data = tile['tile']
             info = zipfile.ZipInfo(tile_name, date_time)
             z.writestr(info, tile_data)
 
-    return [dict(tile=buf.getvalue(), format=zip_format, coord=coord,
+    return [dict(tile=buf.getvalue(), format=zip_format, coord=parent,
                  layer=layer)]
 
 
+def _common_parent(a, b):
+    """
+    Find the common parent tile of both a and b. The common parent is the tile
+    at the highest zoom which both a and b can be transformed into by lowering
+    their zoom levels.
+    """
+
+    if a.zoom < b.zoom:
+        a = a.zoomTo(b.zoom).container()
+
+    elif a.zoom > b.zoom:
+        b = b.zoomTo(a.zoom).container()
+
+    while a.row != b.row or a.column != b.column:
+        a = a.zoomBy(-1).container()
+        b = b.zoomBy(-1).container()
+
+    # by this point a == b.
+    return a
+
+
+def _parent_tile(tiles):
+    """
+    Find the common parent tile for a sequence of tiles.
+    """
+    parent = None
+    for t in tiles:
+        if parent is None:
+            parent = t
+
+        else:
+            parent = _common_parent(parent, t)
+
+    return parent
+
+
 def make_metatiles(size, tiles, date_time=None):
     """
-    Group by coordinates and layers, and make metatiles out of all the tiles
-    which share those properties. Provide a 6-tuple date_time to set the
-    timestamp on each tile within the metatile, or leave it as None to use
-    the current time.
+    Group by layers, and make metatiles out of all the tiles which share those
+    properties relative to the "top level" tile which is parent of them all.
+    Provide a 6-tuple date_time to set the timestamp on each tile within the
+    metatile, or leave it as None to use the current time.
     """
 
     groups = defaultdict(list)
     for tile in tiles:
-        key = (tile['layer'], tile['coord'])
+        key = tile['layer']
         groups[key].append(tile)
 
     metatiles = []
     for group in groups.itervalues():
-        metatiles.extend(make_single_metatile(size, group, date_time))
+        parent = _parent_tile([t['coord'] for t in group])
```

Comment:
Something to think about: with this syntax you're creating a generator (`t['coord'] for t in group`), creating a list by iterating over that generator (`[]`) and then passing the list into the `_parent_tile()` function.

Since the `_parent_tile()` function is simply iterating over the items, you don't need to generate the intermediary list and just pass in the generator:

```python
>>> def foo(foos):
...   print type(foos)
...   for a in foos:
...     print a
...
>>> bar = [dict(a='p', b='c'), dict(a='d', b='e')]
>>> foo([b['a'] for b in bar])
<type 'list'>
p
d
>>> foo(b['a'] for b in bar)
<type 'generator'>
p
d
```

It's probably not worth the change here because the group is always going to be pretty small.